File: bin/otsuge.py
#!/usr/bin/env python3
"""otsuge — おつげ（週ごとの観測）

うらないババが週に1度、拠点を読んで告げる。数の羅列はステータスと年表に
あるので、こちらは**その週に何が起きて、先週と何が違ったか**を書く。

出力は `otsuge/<ISO年>-W<週>.md`（月曜はじまりの ISO 週）。

ふくろ・ステータスと同じく**拠点の中しか見ない**。走らせる順番は
… → teato → fukuro → status → otsuge。

## その週の目でだけ書く

「45日止まっている」のような話は、**その週の終わりの時点**で数える。
今日から数えると、過去のおつげが毎晩書き換わって、読み返したときに
「あのとき何と言われたか」が残らない。未来を知らないおつげは二度と
変わらないので、冪等が完全に保たれる。

掟:
  - 決定論的: 走らせた日で結果が変わらない。
  - 冪等: 内容が変わらなければファイルに触れない。
  - 手で書かせない: 素材が増えれば勝手に増える。

使い方:
    otsuge.py                   # 全部
    otsuge.py --dry-run
    otsuge.py --quiet
"""

from __future__ import annotations

import argparse
import re
import sys
from collections import Counter, defaultdict
from datetime import date, timedelta
from pathlib import Path

from dougu import KYOTEN, frontmatter, read_text, split_frontmatter, write_if_changed

ROOM = KYOTEN / "otsuge"

# ひさしぶりと見なす空き。1か月ぶりに戻ってきたら言う価値がある。
BACK_AFTER = 30

# 止まっていると見なす日数と、それを言うに値する重み。数コミットで
# 終わった実験まで並べると「止まっているもの」が数十行になる。
STALE_AFTER = 60
STALE_MIN_COMMITS = 10
STALE_SHOWN = 8

RE_COMMIT = re.compile(r"^- `([0-9a-f]{4,})` ")
RE_FILE = re.compile(r"^- `([^`]+)`")
RE_HEAD2 = re.compile(r"^## (.+)$")
RE_HEAD3 = re.compile(r"^### (.+)$")

EXT_IGNORE = {
    "png", "jpg", "jpeg", "gif", "svg", "ico", "webp", "avif", "pdf",
    "lock", "sum", "map", "min", "snap", "log", "txt", "csv",
}
ALIAS = {
    "ts": "typescript", "tsx": "typescript", "mts": "typescript",
    "js": "javascript", "mjs": "javascript", "cjs": "javascript", "jsx": "javascript",
    "go": "golang", "py": "python", "rb": "ruby", "rs": "rust",
    "md": "markdown", "sh": "shell", "bash": "shell", "yml": "yaml",
    "scss": "css", "sass": "css",
}


class Day:
    __slots__ = ("commits", "projects", "exts", "troubles", "trouble_tools",
                 "articles", "posts", "utterances", "sessions", "titles")

    def __init__(self) -> None:
        self.commits = 0
        self.projects: Counter = Counter()
        self.exts: Counter = Counter()
        self.troubles = 0
        self.trouble_tools: Counter = Counter()
        self.articles = 0
        self.posts = 0
        self.utterances = 0
        self.sessions = 0
        self.titles: list[str] = []


def ext_name(path: str) -> str:
    base = path.rsplit("/", 1)[-1]
    ext = base.rsplit(".", 1)[-1].lower() if "." in base else base.lower()
    if not ext or len(ext) > 12 or ext in EXT_IGNORE or not ext.isalnum():
        return ""
    return ALIAS.get(ext, ext)


def day_of(days: dict[str, Day], key: str) -> Day:
    if key not in days:
        days[key] = Day()
    return days[key]


def scan(days: dict[str, Day]) -> None:
    """拠点の各部屋を日付ごとに集める。"""
    root = KYOTEN / "teato"
    if root.is_dir():
        for path in sorted(root.rglob("*.md")):
            fields, body, _ = split_frontmatter(read_text(path))
            date_key = fields.get("date", "")
            if not date_key:
                continue
            day = day_of(days, date_key)
            project = section = ""

            for line in body.split("\n"):
                got = RE_HEAD2.match(line)
                if got:
                    project = got.group(1).strip()
                    section = ""
                    continue
                got = RE_HEAD3.match(line)
                if got:
                    section = got.group(1).strip()
                    continue

                if section == "つくった" and RE_COMMIT.match(line):
                    day.commits += 1
                    if project:
                        day.projects[project] += 1
                elif section == "さわった":
                    got = RE_FILE.match(line)
                    if got:
                        name = ext_name(got.group(1))
                        if name:
                            day.exts[name] += 1
                elif section == "つまずいた" and line.startswith("- "):
                    day.troubles += 1
                    # "- 12:34:56 Bash `…`" の道具の名前だけ数える。
                    parts = line.split()
                    if len(parts) >= 3:
                        day.trouble_tools[parts[2]] += 1

    root = KYOTEN / "soto"
    if root.is_dir():
        for path in sorted(root.rglob("*.md")):
            fields, _, _ = split_frontmatter(read_text(path)[:3000])
            date_key = fields.get("date", "")
            if not date_key:
                continue
            day = day_of(days, date_key)
            if fields.get("source") == "polidog.jp":
                day.articles += 1
                if fields.get("title"):
                    day.titles.append(fields["title"])
            else:
                day.posts += 1

    root = KYOTEN / "kotonoha"
    if root.is_dir():
        for path in sorted(root.rglob("*.md")):
            fields, _, _ = split_frontmatter(read_text(path)[:2000])
            date_key = fields.get("date", "")
            if date_key:
                try:
                    day_of(days, date_key).utterances += int(fields.get("utterances", 0))
                except (TypeError, ValueError):
                    pass

    root = KYOTEN / "bouken"
    if root.is_dir():
        for path in sorted(root.rglob("*.md")):
            fields, _, _ = split_frontmatter(read_text(path)[:2000])
            date_key = (fields.get("started") or "")[:10]
            if date_key:
                day_of(days, date_key).sessions += 1


# ---------------------------------------------------------------- 週に畳む

def monday(day: date) -> date:
    return day - timedelta(days=day.weekday())


def week_key(day: date) -> str:
    iso = day.isocalendar()
    return f"{iso[0]}-W{iso[1]:02d}"


def parse(key: str) -> date | None:
    try:
        return date(int(key[:4]), int(key[5:7]), int(key[8:10]))
    except (ValueError, IndexError):
        return None


class Week:
    __slots__ = ("key", "start", "end", "commits", "projects", "exts",
                 "troubles", "trouble_tools", "articles", "posts",
                 "utterances", "sessions", "titles", "days_worked")

    def __init__(self, key: str, start: date) -> None:
        self.key = key
        self.start = start
        self.end = start + timedelta(days=6)
        self.commits = 0
        self.projects: Counter = Counter()
        self.exts: Counter = Counter()
        self.troubles = 0
        self.trouble_tools: Counter = Counter()
        self.articles = 0
        self.posts = 0
        self.utterances = 0
        self.sessions = 0
        self.titles: list[str] = []
        self.days_worked = 0

    def add(self, day: Day) -> None:
        self.commits += day.commits
        self.projects.update(day.projects)
        self.exts.update(day.exts)
        self.troubles += day.troubles
        self.trouble_tools.update(day.trouble_tools)
        self.articles += day.articles
        self.posts += day.posts
        self.utterances += day.utterances
        self.sessions += day.sessions
        self.titles.extend(day.titles)
        if day.commits:
            self.days_worked += 1

    def empty(self) -> bool:
        return not (self.commits or self.articles or self.posts
                    or self.utterances or self.sessions)


def fold(days: dict[str, Day]) -> list[Week]:
    weeks: dict[str, Week] = {}
    for key, day in days.items():
        got = parse(key)
        if not got:
            continue
        start = monday(got)
        week = weeks.setdefault(week_key(got), Week(week_key(got), start))
        week.add(day)
    return [w for _, w in sorted(weeks.items()) if not w.empty()]


# ---------------------------------------------------------------- 告げる

def diff(now: int, before: int | None) -> str:
    if before is None:
        return ""
    delta = now - before
    if delta == 0:
        return f"（先週 {before:,} / 同じ）"
    return f"（先週 {before:,} / {delta:+,}）"


def render(week: Week, before: Week | None, seen: dict[str, str],
           totals: dict[str, int]) -> str:
    """1週ぶんのおつげ。

    `seen` はその週に入る**前**までの、プロジェクトごとの最後の日。
    `totals` は同じくコミットの累計。どちらも未来を含まない。
    """
    head = frontmatter({
        "room": "otsuge",
        "week": week.key,
        "from": week.start.isoformat(),
        "to": week.end.isoformat(),
        "commits": week.commits,
    })

    body = [f"# {week.key} のおつげ", f"{week.start} 〜 {week.end}"]

    lines = []
    if week.commits:
        lines.append(f"- コミット {week.commits:,}"
                     f"{diff(week.commits, before.commits if before else None)}"
                     f"　手を動かした日 {week.days_worked}")
    if week.articles:
        lines.append(f"- 記事 {week.articles:,}"
                     f"{diff(week.articles, before.articles if before else None)}")
    if week.posts:
        lines.append(f"- SNS {week.posts:,} 日ぶん"
                     f"{diff(week.posts, before.posts if before else None)}")
    if week.sessions:
        lines.append(f"- 会話 {week.sessions:,}"
                     f"{diff(week.sessions, before.sessions if before else None)}"
                     + (f"　発言 {week.utterances:,}" if week.utterances else ""))
    if lines:
        body.append("## 今週\n\n" + "\n".join(lines))

    if week.projects:
        lines = []
        for name, n in week.projects.most_common(8):
            mark = ""
            if name not in seen:
                mark = "（はじめて）"
            else:
                gap = (week.start - date.fromisoformat(seen[name])).days
                if gap >= BACK_AFTER:
                    mark = f"（{gap}日ぶり）"
            lines.append(f"- {name} {n:,}{mark}")
        body.append("## よくいた場所\n\n" + "\n".join(lines))

    if week.exts:
        body.append("## 手が動いたもの\n\n" + "、".join(
            f"{name} {n:,}" for name, n in week.exts.most_common(8)))

    if week.troubles:
        detail = "、".join(f"{t} {n}" for t, n in week.trouble_tools.most_common(5))
        body.append(f"## つまずき\n\n{week.troubles} 件"
                    f"{diff(week.troubles, before.troubles if before else None)}"
                    + (f"\n\n{detail}" if detail else ""))

    if week.titles:
        body.append("## そとに出したもの\n\n" + "\n".join(f"- {t}" for t in week.titles))

    # 止まっているもの。**この週の終わりの時点で**数える。
    stale = []
    for name, last in seen.items():
        if name in week.projects:
            continue
        if totals.get(name, 0) < STALE_MIN_COMMITS:
            continue
        gap = (week.end - date.fromisoformat(last)).days
        if gap >= STALE_AFTER:
            stale.append((gap, name, totals.get(name, 0)))
    if stale:
        stale.sort(reverse=True)
        body.append("## 止まっているもの\n\n" + "\n".join(
            f"- {name} {gap}日（{n:,} コミット積んで）"
            for gap, name, n in stale[:STALE_SHOWN]))

    return head + "\n\n" + "\n\n".join(body) + "\n"


def main() -> int:
    ap = argparse.ArgumentParser(description="おつげ — 週ごとの観測")
    ap.add_argument("--dry-run", action="store_true", help="書かずに結果だけ出す")
    ap.add_argument("--quiet", action="store_true", help="1行だけ報告する")
    args = ap.parse_args()

    days: dict[str, Day] = {}
    scan(days)
    weeks = fold(days)

    # 週を古い順に見ながら「その時点までに知っていること」を育てる。
    # 先に全部集めてしまうと、過去のおつげが未来を知ってしまう。
    seen: dict[str, str] = {}
    totals: dict[str, int] = {}
    stats = {"new": 0, "updated": 0, "same": 0}
    before: Week | None = None

    for week in weeks:
        out = ROOM / f"{week.key}.md"
        stats[write_if_changed(out, render(week, before, dict(seen), dict(totals)),
                               args.dry_run)] += 1

        # この週を見終えてから知識を更新する。
        for name, n in week.projects.items():
            seen[name] = week.end.isoformat()
            totals[name] = totals.get(name, 0) + n
        before = week

    total = sum(stats.values())
    if args.quiet:
        print(f"otsuge: {total}週 (new {stats['new']} / upd {stats['updated']} "
              f"/ same {stats['same']})")
    else:
        if args.dry_run:
            print("（書かずに確認）")
        print(f"  おつげ       : {total:,} 週")
        print(f"    あたらしい : {stats['new']:,}")
        print(f"    かきかえ   : {stats['updated']:,}")
        print(f"    かわらず   : {stats['same']:,}")
        if weeks:
            print(f"  期間         : {weeks[0].key} 〜 {weeks[-1].key}")
        print(f"  ばしょ : {ROOM}")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: bin/otsuge.ts
#!/usr/bin/env node
/**
 * otsuge — おつげ（週ごとの観測）
 *
 * うらないババが週に1度、拠点を読んで告げる。数の羅列はステータスと年表に
 * あるので、こちらは**その週に何が起きて、先週と何が違ったか**を書く。
 *
 * 出力は `otsuge/<ISO年>-W<週>.md`（月曜はじまりの ISO 週）。
 *
 * ふくろ・ステータスと同じく**拠点の中しか見ない**。走らせる順番は
 * … → teato → fukuro → status → otsuge。
 *
 * ## その週の目でだけ書く
 *
 * 「45日止まっている」のような話は、**その週の終わりの時点**で数える。
 * 今日から数えると、過去のおつげが毎晩書き換わって、読み返したときに
 * 「あのとき何と言われたか」が残らない。未来を知らないおつげは二度と
 * 変わらないので、冪等が完全に保たれる。
 *
 * 掟:
 *   - 決定論的: 走らせた日で結果が変わらない。
 *   - 冪等: 内容が変わらなければファイルに触れない。
 *   - 手で書かせない: 素材が増えれば勝手に増える。
 *
 * 使い方:
 *     otsuge.ts                   # 全部
 *     otsuge.ts --dry-run
 *     otsuge.ts --quiet
 */

import { existsSync } from "node:fs";
import { join } from "node:path";

import {
  KYOTEN,
  frontmatter,
  n,
  readText,
  splitFrontmatter,
  writeIfChanged,
  type WriteState,
} from "./dougu.ts";
import { listFiles, parseArgs } from "./cli.ts";

const ROOM = join(KYOTEN, "otsuge");

/** ひさしぶりと見なす空き。1か月ぶりに戻ってきたら言う価値がある。 */
const BACK_AFTER = 30;

/**
 * 止まっていると見なす日数と、それを言うに値する重み。数コミットで
 * 終わった実験まで並べると「止まっているもの」が数十行になる。
 */
const STALE_AFTER = 60;
const STALE_MIN_COMMITS = 10;
const STALE_SHOWN = 8;

const RE_COMMIT = /^- `([0-9a-f]{4,})` /;
const RE_FILE = /^- `([^`]+)`/;
const RE_HEAD2 = /^## ([^\n]+)$/;
const RE_HEAD3 = /^### ([^\n]+)$/;

const EXT_IGNORE = new Set([
  "png", "jpg", "jpeg", "gif", "svg", "ico", "webp", "avif", "pdf",
  "lock", "sum", "map", "min", "snap", "log", "txt", "csv",
]);
const ALIAS: Record<string, string> = {
  ts: "typescript", tsx: "typescript", mts: "typescript",
  js: "javascript", mjs: "javascript", cjs: "javascript", jsx: "javascript",
  go: "golang", py: "python", rb: "ruby", rs: "rust",
  md: "markdown", sh: "shell", bash: "shell", yml: "yaml",
  scss: "css", sass: "css",
};

const MS_PER_DAY = 86_400_000;

function bump(counter: Map<string, number>, key: string, by = 1): void {
  counter.set(key, (counter.get(key) ?? 0) + by);
}

function merge(into: Map<string, number>, from: Map<string, number>): void {
  for (const [k, v] of from) bump(into, k, v);
}

function mostCommon(counter: Map<string, number>, limit: number): [string, number][] {
  return [...counter.entries()].sort((a, b) => b[1] - a[1]).slice(0, limit);
}

class Day {
  commits = 0;
  readonly projects = new Map<string, number>();
  readonly exts = new Map<string, number>();
  troubles = 0;
  readonly troubleTools = new Map<string, number>();
  articles = 0;
  posts = 0;
  utterances = 0;
  sessions = 0;
  readonly titles: string[] = [];
}

function extName(path: string): string {
  const base = path.slice(path.lastIndexOf("/") + 1);
  const ext = (base.includes(".") ? base.slice(base.lastIndexOf(".") + 1) : base).toLowerCase();
  if (!ext || ext.length > 12 || EXT_IGNORE.has(ext) || !/^[0-9a-z]+$/i.test(ext)) return "";
  return ALIAS[ext] ?? ext;
}

function dayOf(days: Map<string, Day>, key: string): Day {
  let d = days.get(key);
  if (!d) {
    d = new Day();
    days.set(key, d);
  }
  return d;
}

/** 拠点の各部屋を日付ごとに集める。 */
function scan(days: Map<string, Day>): void {
  const teato = join(KYOTEN, "teato");
  if (existsSync(teato)) {
    for (const path of listFiles(teato, ".md")) {
      const [fields, body] = splitFrontmatter(readText(path));
      const dateKey = fields.date ?? "";
      if (!dateKey) continue;
      const day = dayOf(days, dateKey);
      let project = "";
      let section = "";

      for (const line of body.split("\n")) {
        const gotHead2 = RE_HEAD2.exec(line);
        if (gotHead2) {
          project = gotHead2[1].trim();
          section = "";
          continue;
        }
        const gotHead3 = RE_HEAD3.exec(line);
        if (gotHead3) {
          section = gotHead3[1].trim();
          continue;
        }

        if (section === "つくった" && RE_COMMIT.test(line)) {
          day.commits += 1;
          if (project) bump(day.projects, project);
        } else if (section === "さわった") {
          const got = RE_FILE.exec(line);
          if (got) {
            const name = extName(got[1]);
            if (name) bump(day.exts, name);
          }
        } else if (section === "つまずいた" && line.startsWith("- ")) {
          day.troubles += 1;
          // "- 12:34:56 Bash `…`" の道具の名前だけ数える。
          const parts = line.split(/\s+/);
          if (parts.length >= 3) bump(day.troubleTools, parts[2]);
        }
      }
    }
  }

  const soto = join(KYOTEN, "soto");
  if (existsSync(soto)) {
    for (const path of listFiles(soto, ".md")) {
      const [fields] = splitFrontmatter(readText(path).slice(0, 3000));
      const dateKey = fields.date ?? "";
      if (!dateKey) continue;
      const day = dayOf(days, dateKey);
      if (fields.source === "polidog.jp") {
        day.articles += 1;
        if (fields.title) day.titles.push(fields.title);
      } else {
        day.posts += 1;
      }
    }
  }

  const kotonoha = join(KYOTEN, "kotonoha");
  if (existsSync(kotonoha)) {
    for (const path of listFiles(kotonoha, ".md")) {
      const [fields] = splitFrontmatter(readText(path).slice(0, 2000));
      const dateKey = fields.date ?? "";
      if (!dateKey) continue;
      const got = Number.parseInt(fields.utterances ?? "0", 10);
      if (!Number.isNaN(got)) dayOf(days, dateKey).utterances += got;
    }
  }

  const bouken = join(KYOTEN, "bouken");
  if (existsSync(bouken)) {
    for (const path of listFiles(bouken, ".md")) {
      const [fields] = splitFrontmatter(readText(path).slice(0, 2000));
      const dateKey = (fields.started ?? "").slice(0, 10);
      if (dateKey) dayOf(days, dateKey).sessions += 1;
    }
  }
}

// ---------------------------------------------------------------- 週に畳む

function toUtc(date: string): number {
  return Date.parse(date + "T00:00:00Z");
}

function fromUtc(ms: number): string {
  return new Date(ms).toISOString().slice(0, 10);
}

function monday(date: string): string {
  const ms = toUtc(date);
  const day = new Date(ms).getUTCDay() || 7; // 月曜=1, 日曜=7
  return fromUtc(ms - (day - 1) * MS_PER_DAY);
}

/** Python の `date.isocalendar()` と同じ ISO 週。 */
function weekKey(date: string): string {
  const t = new Date(toUtc(date));
  const day = t.getUTCDay() || 7;
  t.setUTCDate(t.getUTCDate() + 4 - day); // その週の木曜へ
  const year = t.getUTCFullYear();
  const week = Math.ceil(((t.getTime() - Date.UTC(year, 0, 1)) / MS_PER_DAY + 1) / 7);
  return `${year}-W${String(week).padStart(2, "0")}`;
}

class Week {
  readonly key: string;
  readonly start: string;
  readonly end: string;
  commits = 0;
  readonly projects = new Map<string, number>();
  readonly exts = new Map<string, number>();
  troubles = 0;
  readonly troubleTools = new Map<string, number>();
  articles = 0;
  posts = 0;
  utterances = 0;
  sessions = 0;
  readonly titles: string[] = [];
  daysWorked = 0;

  constructor(key: string, start: string) {
    this.key = key;
    this.start = start;
    this.end = fromUtc(toUtc(start) + 6 * MS_PER_DAY);
  }

  add(day: Day): void {
    this.commits += day.commits;
    merge(this.projects, day.projects);
    merge(this.exts, day.exts);
    this.troubles += day.troubles;
    merge(this.troubleTools, day.troubleTools);
    this.articles += day.articles;
    this.posts += day.posts;
    this.utterances += day.utterances;
    this.sessions += day.sessions;
    this.titles.push(...day.titles);
    if (day.commits) this.daysWorked += 1;
  }

  empty(): boolean {
    return !(this.commits || this.articles || this.posts || this.utterances || this.sessions);
  }
}

function fold(days: Map<string, Day>): Week[] {
  const weeks = new Map<string, Week>();
  for (const [key, day] of days) {
    if (Number.isNaN(toUtc(key))) continue;
    const wk = weekKey(key);
    let week = weeks.get(wk);
    if (!week) {
      week = new Week(wk, monday(key));
      weeks.set(wk, week);
    }
    week.add(day);
  }
  return [...weeks.keys()].sort().map((k) => weeks.get(k)!).filter((w) => !w.empty());
}

// ---------------------------------------------------------------- 告げる

function diff(now: number, before: number | null): string {
  if (before === null) return "";
  const delta = now - before;
  if (delta === 0) return `（先週 ${n(before)} / 同じ）`;
  return `（先週 ${n(before)} / ${delta > 0 ? "+" : ""}${n(delta)}）`;
}

/**
 * 1週ぶんのおつげ。
 *
 * `seen` はその週に入る**前**までの、プロジェクトごとの最後の日。
 * `totals` は同じくコミットの累計。どちらも未来を含まない。
 */
function render(
  week: Week,
  before: Week | null,
  seen: ReadonlyMap<string, string>,
  totals: ReadonlyMap<string, number>,
): string {
  const head = frontmatter({
    room: "otsuge",
    week: week.key,
    from: week.start,
    to: week.end,
    commits: week.commits,
  });

  const body: string[] = [`# ${week.key} のおつげ`, `${week.start} 〜 ${week.end}`];

  const lines: string[] = [];
  if (week.commits) {
    lines.push(`- コミット ${n(week.commits)}${diff(week.commits, before?.commits ?? null)}` +
      `　手を動かした日 ${week.daysWorked}`);
  }
  if (week.articles) {
    lines.push(`- 記事 ${n(week.articles)}${diff(week.articles, before?.articles ?? null)}`);
  }
  if (week.posts) {
    lines.push(`- SNS ${n(week.posts)} 日ぶん${diff(week.posts, before?.posts ?? null)}`);
  }
  if (week.sessions) {
    lines.push(`- 会話 ${n(week.sessions)}${diff(week.sessions, before?.sessions ?? null)}` +
      (week.utterances ? `　発言 ${n(week.utterances)}` : ""));
  }
  if (lines.length) body.push("## 今週\n\n" + lines.join("\n"));

  if (week.projects.size) {
    const rows: string[] = [];
    for (const [name, count] of mostCommon(week.projects, 8)) {
      let mark = "";
      const last = seen.get(name);
      if (last === undefined) {
        mark = "（はじめて）";
      } else {
        const gap = Math.round((toUtc(week.start) - toUtc(last)) / MS_PER_DAY);
        if (gap >= BACK_AFTER) mark = `（${gap}日ぶり）`;
      }
      rows.push(`- ${name} ${n(count)}${mark}`);
    }
    body.push("## よくいた場所\n\n" + rows.join("\n"));
  }

  if (week.exts.size) {
    body.push("## 手が動いたもの\n\n" +
      mostCommon(week.exts, 8).map(([name, count]) => `${name} ${n(count)}`).join("、"));
  }

  if (week.troubles) {
    const detail = mostCommon(week.troubleTools, 5).map(([t, c]) => `${t} ${c}`).join("、");
    body.push(`## つまずき\n\n${week.troubles} 件${diff(week.troubles, before?.troubles ?? null)}` +
      (detail ? `\n\n${detail}` : ""));
  }

  if (week.titles.length) {
    body.push("## そとに出したもの\n\n" + week.titles.map((t) => `- ${t}`).join("\n"));
  }

  // 止まっているもの。**この週の終わりの時点で**数える。
  const stale: [number, string, number][] = [];
  for (const [name, last] of seen) {
    if (week.projects.has(name)) continue;
    const total = totals.get(name) ?? 0;
    if (total < STALE_MIN_COMMITS) continue;
    const gap = Math.round((toUtc(week.end) - toUtc(last)) / MS_PER_DAY);
    if (gap >= STALE_AFTER) stale.push([gap, name, total]);
  }
  if (stale.length) {
    // Python の sorted(reverse=True) はタプルの辞書順。
    stale.sort((a, b) => (b[0] - a[0]) || (a[1] < b[1] ? 1 : a[1] > b[1] ? -1 : b[2] - a[2]));
    body.push("## 止まっているもの\n\n" + stale.slice(0, STALE_SHOWN)
      .map(([gap, name, count]) => `- ${name} ${gap}日（${n(count)} コミット積んで）`).join("\n"));
  }

  return head + "\n\n" + body.join("\n\n") + "\n";
}

function main(): number {
  const args = parseArgs(process.argv.slice(2), ["dry-run", "quiet"]);

  const days = new Map<string, Day>();
  scan(days);
  const weeks = fold(days);

  // 週を古い順に見ながら「その時点までに知っていること」を育てる。
  // 先に全部集めてしまうと、過去のおつげが未来を知ってしまう。
  const seen = new Map<string, string>();
  const totals = new Map<string, number>();
  const stats: Record<WriteState, number> = { new: 0, updated: 0, same: 0 };
  let before: Week | null = null;

  for (const week of weeks) {
    const out = join(ROOM, `${week.key}.md`);
    stats[writeIfChanged(out, render(week, before, new Map(seen), new Map(totals)),
      args.flags["dry-run"])] += 1;

    // この週を見終えてから知識を更新する。
    for (const [name, count] of week.projects) {
      seen.set(name, week.end);
      totals.set(name, (totals.get(name) ?? 0) + count);
    }
    before = week;
  }

  const total = stats.new + stats.updated + stats.same;
  if (args.flags.quiet) {
    console.log(
      `otsuge: ${total}週 (new ${stats.new} / upd ${stats.updated} / same ${stats.same})`,
    );
  } else {
    if (args.flags["dry-run"]) console.log("（書かずに確認）");
    console.log(`  おつげ       : ${n(total)} 週`);
    console.log(`    あたらしい : ${n(stats.new)}`);
    console.log(`    かきかえ   : ${n(stats.updated)}`);
    console.log(`    かわらず   : ${n(stats.same)}`);
    if (weeks.length) console.log(`  期間         : ${weeks[0].key} 〜 ${weeks[weeks.length - 1].key}`);
    console.log(`  ばしょ : ${ROOM}`);
  }

  return 0;
}

process.exit(main());
